### DISCUSSION TITLE
Multi cursor/selection editing

### DISCUSSION THREAD

**Discussion by pyscripter** [pyscripter/SynEdit#69] [Ideas]
IMO, one of the most desirable missing features of SynEdit is #2.  In case you are not familiar with multi cursor/selection, the following two links will give you a good idea:

- [VS Code multi cursor/selection editing](https://code.visualstudio.com/docs/editor/codebasics)
- [ATSynEdittmulti cursor/selection editing](https://wiki.freepascal.org/ATSynEdit#Multi-carets)

What SynEdit offers instead is the column mode, which, although useful, is significantly underpowered compare to the above.
My proposal is to replace the column/line modes with proper multi cursor/selection editing.   Preliminary work has been already done, and I expect to make good progress in the near future.

@MShark67 @JaFi-cz @groehner @vincentparrett  Any thoughts/feedback

**Comment by JaFi-cz** [comment]
Good decision.
Build it step by step - colum editing as first step,  followed by multiselection editing (find all occurrences end edit them at once)

**Comment by pyscripter** [comment]
There has been significant progress in the multicaret branch.   The aim is to match the related features of VS-Code.  The hard part is done. The following is meant to help track the progress made.

- [x] Custom caret rendering and blinking
- [x] Multi-selection using the mouse
- [x] Multi-caret command processing
- [x] Undo/redo of multi-caret commands
- [x]  Column selection using the mouse
- [x] Caret and selection creation commands
  - [x] Column select Left, Right, Up, Down, PgUp, PgDown
  - [x] Select next matching word
  - [x] Select all matching words
  - [x] Carets at line ends
- [ ] Other things to be fixed
  - [x] SelText
  - [x]  Multi-caret copy/paste
  - [ ] Printing of selection(s)
  - [x] Search and replace in selection(s)

**Reply by pyscripter** [reply]
I will leave printing of multiple selections for another day.  It will work with the active selection only for now.

Multi-caret editing is now feature complete.   But it needs testing and testing and testing to iron out bugs.  Please help.

**Comment by MShark67** [comment]
Looks amazing!  I haven't used multi-caret very much (besides some of the Delphi editor options.)  I found the multi-caret mouse selection to work intuitively and well.   I also tested some of the undo/redo and it looks great.  I'm not sure exactly how keyboard multi-caret movement is supposed to work (particularly in how to end the multi-caret mode without clicking) but I'm happy to test or help in any way I can.   Great job!

**Reply by pyscripter** [reply]
> particularly in how to end the multi-caret mode without clicking

Press Esc

**Reply by pyscripter** [reply]
> I'm not sure exactly how keyboard multi-caret movement is supposed to work

It's coming soon...

**Reply by MShark67** [reply]
Aha!   Thank you!  It turns out that my personal editor project which I use for testing didn't have the AddKey() call for ecEscape.  Now it works!

**Comment by pyscripter** [comment]
Column selection using the mouse implemented.  Shift+Alt+mouse down/move.
Tested typing with a few thousand cursors and it is still quite fast, without doing any optimizations.  For comparison VS-code has a limit of 10000 cursors, but gets very slow well before reaching that limit.

**Comment by pyscripter** [comment]
Key column selection commands implemented:
```pascal
  AddKey(ecSelColumnLeft, SYNEDIT_LEFT, [ssShift, ssAlt]);
  AddKey(ecSelColumnRight, SYNEDIT_RIGHT, [ssShift, ssAlt]);
  AddKey(ecSelColumnUp, SYNEDIT_UP, [ssShift, ssAlt]);
  AddKey(ecSelColumnDown, SYNEDIT_DOWN, [ssShift, ssAlt]);
  AddKey(ecSelColumnPageUp, SYNEDIT_PRIOR, [ssShift, ssAlt]);
  AddKey(ecSelColumnPageDown, SYNEDIT_NEXT, [ssShift, ssAlt]);
```

**Comment by pyscripter** [comment]
Implemented adding to selections the next matching word and selection of all matching words.

Select Word at Cursor:  Shortcut Ctrl+W.   If you press it again it adds to selection the next match.  Press it again and adds the next match etc.

Select Matching Text:   Shortcut Shift+Ctrl+W  Selects all text matching the current selection.  If there is no selection it selects all occurrences of the word at cursor.

The above provide a way to achieve the "Sync Edit" functionality of Delphi.

**Comment by pyscripter** [comment]
Video Tutorial showcasing the multi-caret, multi-selection capabilities of SynEdit.  (They are using PyScripter which is based on SynEdit).  
@MShark67 @JaFi-cz @dados @vincentparrett @groehner @jonjbar @m-matsubara
Please try and report/help fix any issues you find.

**Comment by pyscripter** [comment]
I have added [multi-caret programming and compatibility notes](https://github.com/pyscripter/SynEdit/blob/multicaret/Doc/Multi-caret.md).

**Comment by MShark67** [comment]
Looks and works great!  Nice documentation too, that really helps.   I did notice one problem.   If you make multiple carets and then unfocus the editor (by changing applications or opening a modeless form and switching to it) then if you scroll the editor using the mousewheel then the carets appearance becomes unsynchronized from the scrolling (it looks like the caret positions aren't moving with the text in that case.)  Once you refocus the editor and cause a refresh it seems to fix itself.   I've also seen it happen when programmatically moving the caret and selection (using a modeless search and replace type feature.)    Let me know if you need more details!

**Reply by pyscripter** [reply]
Could you please try with the latest commits and report back if the issue is fixed.

Thanks for testing and finding issues.   Hit it hard!

**Reply by MShark67** [reply]
That issue is fixed!  Nice work.  I do notice something else.  I see a lot of paint artifacts and scrolling while doing a massive mult-selection replace.  I'm stress testing this by loading System.IniFiles.pas (this just happened to be a file I was looking at recently)
Steps:
1.  Load System.IniFiles.pas.
2. Select the word "string"
3. Press Ctrl-Shift-W to select all "string" throughout the document (it's a lot!)
4. Start typing.
5. Notice that there's a lot of redrawing of the screen and the scroll position seems to change.  When the work is done all artifacts are gone but the scroll position has changed such that the first "string" selection is now at the topline of the editor (which may be "as intended."  I assume it's some call to "make caret location visible" somewhere.)

Let me know if you can't replicate this as it could be something I'm doing wrong in my test app.

**Reply by pyscripter** [reply]
Could you please try again with the latest commit and report back?

**Reply by MShark67** [reply]
Works great!  I've noticed that if you have a painttransient handler than you get all kinds of odd artefacts (I do my own bracket highlighting in my editor for some reason.)  I think it's being called many times, and maybe shouldn't be if paintlock is on (or some other solution, that's a guess!)

**Reply by pyscripter** [reply]
Transient painting does not blend well with the D2D drawing and should be avoided if possible using indicators.  Built-in spellchecking and bracket highlighting do this and can serve as examples of how to do it.  I am keeping transient painting for backward compatibility.

In the latest commit, I tried to rationalize transient painting.
- I have replaced DoOnPaintTransient with DoOnPaintTransientEx and removed the latter.    It does not make sense to have nested calls to transient painting handlers.
- I have removed many calls to DoOnPaintTransient and left 7 pairs. 
- Transient painting only happens if FPaintLock = 0

Since I am not using transient painting handlers, could you please review and test the changes?  If it is still not working well, could you please suggest changes to make it work?

**Reply by MShark67** [reply]
Looks great!  Totally solves the issues I was seeing.  I'll look into replacing my use of PaintTransient, but your changes seem to get everything working well for me.  Excellent stuff!

**Reply by pyscripter** [reply]
Thanks for your efforts. Please keep trying to discover problem areas.

**Comment by MShark67** [comment]
Got a chance to do some testing of all the new features including multi-caret editing, zoom, selection caret-arrow changes (this is a small but really nice change btw), new structure highlighting, and to a lesser extent, the annotated scrollbar feature.   I didn't notice any problems at all.  Amazing work and I think it's mergeable at this stage (which I think will bring much more feedback.)

**Reply by pyscripter** [reply]
> selection caret-arrow changes (this is a small but really nice change btw)

Thanks Mark! I am not sure what you referring to.  Could you please explain?

**Reply by MShark67** [reply]
Oops, sorry.  Was referring to the change for issue #100.  Thanks!

**Comment by pyscripter** [comment]
The multicaret branch was merged back to master and deleted.

**Comment by MShark67** [comment]
The multi-caret stuff is fantastic!  One feature that I'd like to see if possible is allowing an initial selection combined with Ctrl-W (ecSelWord) to then add a new selection of the next occurrence of the selected text.  Basically, the same as the current Ctrl-W behavior but allowing an initial and possibly non-word-oriented selection.  This would match VSCode's Ctrl-D behavior when a selection already exists.  Btw, are feature requests considered "issues" with GitHub?  I wasn't sure a good place to put this.  Thanks, as always!

**Reply by pyscripter** [reply]
> Btw, are feature requests considered "issues" with GitHub?

Yes

If I understand well This is what you suggest:

Ctrl + W
No selection:  Select Word
Selection:  Add cursor to the next match **whether the selection is a word or not**

Shift + Ctrl + W
No selection:  Select Word at cursor and add cursors to all matches
Selection:  Add cursors to all matches whether the selection is a word or not

The only difference with current behavior is the bolded part.  in Ctrl+W with selection.

Could you please confirm?

**Reply by MShark67** [reply]
Yes, the bolded part is exactly what I meant.  Thanks!

**Reply by pyscripter** [reply]
Could you please open a new issue to keep track of this?

**Comment by dados** [comment]
Have you guys noticed that Shift+Alt for column selection is the default combination in Windows for next type language (if you have multiple languages set up) I had Icelandic as primary and I also had UK English and US English
I thought I was going mad when some keys where coming out differently hahahaha
I just removed the English language typing option from Windows, but it might be an issue for some....

**Reply by pyscripter** [reply]
The language only change when you press Alt+Shift and release them.  It does not change when say you press Alt+Shift and Click or Alt+Shift + arrow.  No need to remove the Windows option.  Just takes a bit of time to get used to it.

VS Code and other editors use the same key combination.

Having said that, I also sometimes accidentally change the language...
